src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmCodeGenerationServiceImpl.java

```diff
@@ -0,0 +1,202 @@
+
+		final List<String> executionOrder = Arrays.asList(this.ruleFactory.getRuleSequenceForNamespace(CODING_RULE_SEQUENCE));
+		final CodingRuleExecutionContext codingRuleExecutionContext = new CodingRuleExecutionContext(executionOrder, namingConfiguration);
+		final List<GermplasmCodingResult> assignCodesResultsList = new ArrayList<>();
+
+		for (final Integer gid : gidsToProcess) {
+			final String generatedCodeName = (String) this.rulesService.runRules(codingRuleExecutionContext);
+			assignCodesResultsList.add(this.applyGroupName(gid, nameType, generatedCodeName));
+			codingRuleExecutionContext.reset();
+		}
+
+		return assignCodesResultsList;
+	}
+
+	protected List<GermplasmCodingResult> applyGroupNamesForManualNaming(final Set<Integer> gids, final GermplasmNameSetting setting,
+		final GermplasmNameTypeDTO nameType) throws InvalidGermplasmNameSettingException {
+		final List<GermplasmCodingResult> assignCodesResultsList = new ArrayList<>();
+		final boolean startNumberSpecified = setting.getStartNumber() != null;
+		Integer startNumber = setting.getStartNumber();
+
+		// Call this method to check first if the name settings are valid.
+		this.germplasmNamingService.getNextNameInSequence(setting);
+
+		for (final Integer gid : gids) {
+			// Increment start number of succeeding germplasm processed based on initial start # specified, if any
+			if (startNumberSpecified) {
+				setting.setStartNumber(startNumber++);
+			}
+			final String nameWithSequence = this.germplasmNamingService.generateNextNameAndIncrementSequence(setting);
+			final GermplasmCodingResult result = this.applyGroupName(gid, nameType, nameWithSequence);
+			assignCodesResultsList.add(result);
+		}
+		return assignCodesResultsList;
+	}
+
+	protected GermplasmCodingResult applyGroupName(final Integer gid,
+		final GermplasmNameTypeDTO nameType, final String generatedCodeName) {
+
+		final GermplasmCodingResult result = new GermplasmCodingResult();
+		result.setGid(gid);
+
+		final Germplasm germplasm = this.germplasmService.getGermplasmByGID(gid);
+
+		if (germplasm.getMgid() == null || germplasm.getMgid() == 0) {
+			result.addMessage(String.format(GERMPLASM_NOT_PART_OF_MANAGEMENT_GROUP, germplasm.getGid()));
+			return result;
+		}
+
+		final List<Germplasm> groupMembers =
+			this.germplasmGroupingService.getDescendantGroupMembers(germplasm.getGid(), germplasm.getMgid());
+		groupMembers.add(0, germplasm);
+
+		for (final Germplasm member : groupMembers) {
+			this.addName(member, generatedCodeName, nameType, result);
+		}
+
+		return result;
+
+	}
+
+	private void addName(final Germplasm germplasm, final String groupName, final GermplasmNameTypeDTO nameType,
+		final GermplasmCodingResult result) {
+
+		final List<Name> currentNames = germplasm.getNames();
+
+		Name existingNameOfGivenType = null;
+		if (!currentNames.isEmpty() && nameType != null) {
+			for (final Name name : currentNames) {
+				if (nameType.getId().equals(name.getTypeId())) {
```

Comment:
You can use code instead if using GermplasmDto

Reply:
Done